<launch>
    <arg name="map_file" />
    <arg name="clean_goal" />

    <!-- AMCL -->
    <node pkg="amcl" type="amcl" name="amcl" output="screen" >
      <param name="use_map_topic"             value="true"/>
      <param name="first_map_only"            value="false"/>
      <param name="odom_model_type"           value="diff-corrected"/>
      <param name="gui_publish_rate"          value="5.0"/>
      <param name="save_pose_rate"            value="1.0"/>
      <param name="laser_max_beams"           value="50"/>
      <param name="laser_min_range"           value="-1.0"/>
      <param name="laser_max_range"           value="-1.0"/>
      <param name="min_particles"             value="500"/>
      <param name="max_particles"             value="5000"/>
      <param name="kld_err"                   value="0.05"/>
      <param name="kld_z"                     value="0.99"/>
      <param name="odom_alpha1"               value="0.02"/>
      <param name="odom_alpha2"               value="0.01"/>
      <param name="odom_alpha3"               value="0.01"/>
      <param name="odom_alpha4"               value="0.02"/>
      <param name="laser_z_hit"               value="0.5"/>
      <param name="laser_z_short"             value="0.05"/>
      <param name="laser_z_max"               value="0.05"/>
      <param name="laser_z_rand"              value="0.5"/>
      <param name="laser_sigma_hit"           value="0.2"/>
      <param name="laser_lambda_short"        value="0.1"/>
      <param name="laser_model_type"          value="likelihood_field"/>
      <param name="laser_likelihood_max_dist" value="2.0"/>
      <param name="update_min_d"              value="0.001"/>
      <param name="update_min_a"              value="0.001"/>
      <param name="odom_frame_id"             value="odom"/>
      <param name="base_frame_id"             value="base_footprint"/>
      <param name="global_frame_id"           value="map"/>
      <param name="resample_interval"         value="1"/>
      <param name="transform_tolerance"       value="0.2"/>
      <param name="recovery_alpha_slow"       value="0.001"/>
      <param name="recovery_alpha_fast"       value="0.1"/>
      <param name="initial_pose_x"            value="0.0"/>
      <param name="initial_pose_y"            value="0.0"/>
      <param name="initial_pose_a"            value="0.0"/>
      <remap from="scan"                      to="/scan_multi"/>
      <remap from="map"                       to="/map"/>
      <remap from="static_map"                to="/static_map"/>
    </node>

    <!--We need a map to fully cover-->
    <node name="grid_server" pkg="map_server" type="map_server" args="$(arg map_file)" output="screen">
        <param name="frame_id" value="map"/>
    </node>

    <!-- Cleaning map -->
    <node name="cleaned_area" pkg="costmap_2d" type="costmap_2d_node" output="screen">
        <rosparam file="$(find rm_cleaner_76912)/config/costmap_clean_params.yaml" command="load" ns="costmap" />
    </node>

    <!-- Cleaning node -->
    <node name="clean_bot" pkg="rm_cleaner_76912" type="cleaner_bot.py" output="screen">
        <param name="map_topic" value="/map"/>
        <param name="cleaned_area_topic" value="/cleaned_area/costmap/costmap" />
        <param name="move_base_ns" value="move_base" />
        <param name="clean_goal" value="$(arg clean_goal)" />
        <param name="delta_w" value="1.0" />
        <param name="delta_h" value="0.7" />
        <param name="frequency" value="15" />
    </node>

    <!-- Initiate global localization, wherein all particles are dispersed randomly through the free space in the map -->
    <!-- <node pkg="rosservice" type="rosservice" name="global_localization" args="call --><!--wait /global_localization"/> -->

</launch>
